{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ae74cce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from augur.utils import json_to_tree\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import MutableSeq\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9466979",
   "metadata": {},
   "source": [
    "### Count mutations in each gene at each node in the tree\n",
    "\n",
    "Strategy: \n",
    "1. put json tree in BioPhylo format to traverse it and count mutations at each node (relative to root)\n",
    "2. save the mutations at each node in a dictionary with the node name as key\n",
    "3. traverse the tree in JSON format and add the mutation information to the JSON object by looking up the node from the above dictionary\n",
    "4. replace the json['tree'] with this edited tree and save the whole json again\n",
    "5. also give the ability to add a color_by option for auspice to color the tree by the mutation tally"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c960fd",
   "metadata": {},
   "source": [
    "First, get the locations of all genes, and find the codons within the gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f683bd64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# store gene location information\n",
    "location_by_gene = {}\n",
    "# store the wuhan-1 sequence for each gene\n",
    "wuhan_seq_by_gene = {}\n",
    "\n",
    "with open(\"reference_all_genes.gb\") as reference_handle:\n",
    "    for record in SeqIO.parse(reference_handle, \"genbank\"):\n",
    "        wuhan_seq = record.seq\n",
    "        for feature in record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                gene = feature.qualifiers['gene'][0]\n",
    "                location = feature.location\n",
    "                location_by_gene[gene] = location\n",
    "                wuhan_seq_by_gene[gene] = feature.location.extract(wuhan_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c90a6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes = list(location_by_gene.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "085b773b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# map nucleotide position to gene\n",
    "# allow a position to map to multiple genes (ex: this will happen for S1 and S, orf9b and N)\n",
    "# if position in not in any of the annotated genes, call it 'noncoding'\n",
    "nt_to_gene_mapper = {i+1:[] for i in range(len(wuhan_seq))}\n",
    "\n",
    "for gene, loc in location_by_gene.items():\n",
    "    for pos in range(loc.start+1, loc.end+1):\n",
    "        nt_to_gene_mapper[pos].append(gene)\n",
    "\n",
    "for k, v in nt_to_gene_mapper.items():\n",
    "    if len(v)==0:\n",
    "        nt_to_gene_mapper[k] = ['noncoding']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "960b755e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_tree(date, tree_type):\n",
    "    \"\"\"\n",
    "    Read in the 2m tree json for the specified date\n",
    "    \"\"\"\n",
    "\n",
    "    # path to tree json\n",
    "    if tree_type == 'sars2':\n",
    "        tree_file = f'../ncov_builds/auspice_{tree_type}_2m/sars2_{date}_2m.json'\n",
    "    elif tree_type == '21L':\n",
    "        tree_file = f'../ncov_builds/auspice_{tree_type}_2m/sars2_21L_{date}_2m.json'    \n",
    "    elif tree_type == 'alltime':\n",
    "        tree_file = f'../ncov_builds/auspice_sars2_{tree_type}/sars2_global_all-time.json'\n",
    "\n",
    "    with open(tree_file, 'r') as f:\n",
    "        tree_json = json.load(f)\n",
    "\n",
    "    # put tree in Bio.phylo format\n",
    "    tree = json_to_tree(tree_json)\n",
    "    \n",
    "    return tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ad2bd78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_mutations_by_type(gene, muts):\n",
    "    \"\"\"\n",
    "    Given a list of mutations in this gene, apply them to the reference sequence \n",
    "    for this gene and count the number of mutations\n",
    "    \"\"\"\n",
    "    # get the gene sequence of wuhan-1 \n",
    "    wuhan_gene_seq = wuhan_seq_by_gene[gene]\n",
    "    translated_wuhan_seq = wuhan_gene_seq.translate()\n",
    "        \n",
    "    \n",
    "    # get the genome coordinates for this gene\n",
    "    gene_start, gene_end = location_by_gene[gene].start, location_by_gene[gene].end\n",
    "    \n",
    "    # apply each mutation in this gene to the wuhan-1 reference sequence\n",
    "    mutated_gene_seq = MutableSeq(wuhan_gene_seq)\n",
    "\n",
    "    \n",
    "    # muts are listed oldest to newest, so if multiple mutations occur at same nt, \n",
    "    # the newer one should correctly overwrite the older one\n",
    "    for mut in muts:\n",
    "        # convert genome coordinates to gene-specific coordinates \n",
    "        mut_pos = int(mut[1:-1])-gene_start-1\n",
    "        # deal with ribosomal frameshift in Nsp12\n",
    "        if gene=='Nsp12' and int(mut[1:-1])>13468:\n",
    "            mut_pos = mut_pos+1\n",
    "\n",
    "        #apply the mutation\n",
    "        mutated_gene_seq[mut_pos] = mut[-1]\n",
    "        # if the mutation happens at the nt where ribosomal frameshift occurs, need to apply the mutation to 2 nt positions\n",
    "        if int(mut[1:-1])==13468:\n",
    "            mutated_gene_seq[mut_pos+1] = mut[-1]\n",
    "\n",
    "        \n",
    "    translated_mut_seq = mutated_gene_seq.translate()\n",
    "\n",
    "    \n",
    "    # now tally the number of each kind of mutation\n",
    "    # find stops (other than those that occur at the very end)\n",
    "    stops_in_gene = str(translated_mut_seq[:-1]).count('*')\n",
    "    \n",
    "    # find number of nonsyn mutations\n",
    "    nonsyn_in_gene = sum(1 for a, b in zip(translated_wuhan_seq, translated_mut_seq) if a != b) - stops_in_gene\n",
    "    \n",
    "\n",
    "    \n",
    "    # find number of syn mutations (as the total number of muts that are not stop or nonsyn)\n",
    "    syn_in_gene = len(muts) - nonsyn_in_gene - stops_in_gene\n",
    "    \n",
    "    return nonsyn_in_gene, syn_in_gene, stops_in_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "086ceeac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_mutations(date, tree_type):\n",
    "    \"\"\"\n",
    "    For each node in the tree, find the number of mutations (syn and nonsyn) at each node \n",
    "    and add these tallies as an attribute of the node\n",
    "    Return a dictionary where key is node name and value is the mutation accumulation info\n",
    "    \"\"\"\n",
    "    tree = readin_tree(date, tree_type)\n",
    "    \n",
    "    # initialize dictionary to store mut info\n",
    "    muts_by_node = {}\n",
    "    \n",
    "    for node in tree.find_clades():\n",
    "        \n",
    "        # keep track of nonsyn, syn, stop muts on this path\n",
    "        # by gene\n",
    "        nonsyn_total_by_gene = {x:0 for x in all_genes}\n",
    "        syn_total_by_gene = {x:0 for x in all_genes}\n",
    "        stop_total_by_gene = {x:0 for x in all_genes}\n",
    "        \n",
    "        # get path back to the root\n",
    "        path = tree.get_path(node)\n",
    "        \n",
    "        # get all nucleotide mutations relative to root\n",
    "        nt_muts = [branch.branch_attrs['mutations'].get('nuc', []) for branch in path]\n",
    "        # flatten the list of nucleotide mutations\n",
    "        nt_muts = [item for sublist in nt_muts for item in sublist]\n",
    "        # ignore deletions and \"reversions\" from deletions\n",
    "        nt_muts = [x for x in nt_muts if '-' not in x] \n",
    "\n",
    "        # find which gene each mut is in\n",
    "        mut_and_genelocation = {mut: nt_to_gene_mapper[int(mut[1:-1])] for mut in nt_muts}\n",
    "        # group the mutations by gene\n",
    "        muts_per_gene = {}\n",
    "        for m, gs in mut_and_genelocation.items():\n",
    "            # some mutations will map to multiple genes\n",
    "            for g in gs:\n",
    "                # don't need to keep track of the noncoding muts\n",
    "                if g!='noncoding':\n",
    "                    if g in muts_per_gene.keys():\n",
    "                        muts_per_gene[g].append(m)\n",
    "                    else:\n",
    "                        muts_per_gene[g] = [m]\n",
    "\n",
    "        # for each gene, find how many syn, and nonsyn and stop muts\n",
    "        for gene, gene_nt_muts in muts_per_gene.items():\n",
    "            nonsyn_in_gene, syn_in_gene, stops_in_gene = count_mutations_by_type(gene, gene_nt_muts)\n",
    "            nonsyn_total_by_gene[gene] = nonsyn_in_gene\n",
    "            syn_total_by_gene[gene] = syn_in_gene\n",
    "            stop_total_by_gene[gene] = stops_in_gene\n",
    "            \n",
    "        node.node_attrs['Nonsyn_muts'] = nonsyn_total_by_gene\n",
    "        node.node_attrs['Syn_muts'] = syn_total_by_gene\n",
    "        node.node_attrs['Stop_muts'] = stop_total_by_gene\n",
    "        \n",
    "        muts_by_node[node.name] = {'Nonsyn_muts':nonsyn_total_by_gene, 'Syn_muts': syn_total_by_gene, \n",
    "                                   'Stop_muts': stop_total_by_gene}\n",
    "    \n",
    "    \n",
    "    return muts_by_node "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "013430c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def traverse_tree(branch, muts_by_node, color_by_genes):\n",
    "    \"\"\"\n",
    "    Traverse the tree and add the mutation accumulation at each node\n",
    "    \"\"\"\n",
    "\n",
    "    branch['mut_accumulation'] = muts_by_node[branch['name']]\n",
    "    \n",
    "    if color_by_genes:\n",
    "        for gene in color_by_genes:\n",
    "            branch['node_attrs'][f'{gene}_mutations'] = {\"value\": muts_by_node[branch['name']]['Nonsyn_muts'][gene]}\n",
    "    \n",
    "    if 'children' in branch.keys():\n",
    "        for child in branch['children']:\n",
    "            traverse_tree(child, muts_by_node, color_by_genes)\n",
    "            \n",
    "    return branch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2ad712b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def edit_tree(date, tree_type, color_by_genes=False):\n",
    "    \"\"\"\n",
    "    Edit the tree as a JSON object to have the mutation tally information at each node\n",
    "    Can also supply a list of genes so that the auspice tree can be colored by mutation tally in these genes\n",
    "    Save the edited tree\n",
    "    tree_type options= ['21L', 'sars2']\n",
    "    \"\"\"\n",
    "    # get dicionary of the mutation tally at each node\n",
    "    muts_by_node = count_mutations(date, tree_type)\n",
    "    \n",
    "    # path to tree json\n",
    "    if tree_type == 'sars2':\n",
    "        tree_file = f'../ncov_builds/auspice_{tree_type}_2m/sars2_{date}_2m.json'\n",
    "    elif tree_type == '21L':\n",
    "        tree_file = f'../ncov_builds/auspice_{tree_type}_2m/sars2_21L_{date}_2m.json'\n",
    "    if tree_type == 'alltime':\n",
    "        tree_file = f'../ncov_builds/auspice_sars2_{tree_type}/sars2_global_all-time.json'\n",
    "\n",
    "    with open(tree_file, 'r') as f:\n",
    "        tree_json = json.load(f)\n",
    "    \n",
    "    # add mutation tallies to each node\n",
    "    tree_edited = traverse_tree(tree_json['tree'], muts_by_node, color_by_genes)\n",
    "    # replace the 'tree' in the auspice json with the edited tree\n",
    "    tree_json['tree'] = tree_edited\n",
    "    \n",
    "    # add color by option for the specified genes\n",
    "    if color_by_genes:\n",
    "        for gene in color_by_genes:\n",
    "            tree_json['meta']['colorings'].append({\n",
    "                \"key\": f\"{gene}_mutations\",\n",
    "                \"title\": f\"{gene} mutations\",\n",
    "                \"type\": \"continuous\"})\n",
    "\n",
    "    # serialize json\n",
    "    json_object = json.dumps(tree_json, indent=2)\n",
    "\n",
    "    # Write out edited tree file\n",
    "    if tree_type == 'sars2':\n",
    "        with open(f'trees_w_mut_counts/{tree_type}_{date}_2m.json', \"w\") as outfile:\n",
    "            outfile.write(json_object)\n",
    "    elif tree_type == '21L':\n",
    "        with open(f'trees_w_mut_counts/sars2_{tree_type}_{date}_2m.json', \"w\") as outfile:\n",
    "            outfile.write(json_object)\n",
    "    elif tree_type == 'alltime':\n",
    "        with open(f'trees_w_mut_counts/sars2_{tree_type}_2023-06.json', \"w\") as outfile:\n",
    "            outfile.write(json_object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7235ac54",
   "metadata": {},
   "outputs": [],
   "source": [
    "edit_tree('2023-06', 'alltime', color_by_genes=all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bce6c968",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edit all 2m trees\n",
    "all_dates = ['2020-03', '2020-05', '2020-07', '2020-09', '2020-11', \n",
    "             '2021-01', '2021-03', '2021-05', '2021-07', '2021-09', '2021-11', \n",
    "             '2022-01', '2022-03', '2022-05', '2022-07', '2022-09', '2022-11', \n",
    "             '2023-01', '2023-03', '2023-05']\n",
    "\n",
    "for d in all_dates:\n",
    "    edit_tree(d, 'sars2', color_by_genes=all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "15bbf970",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edit all 2m 21L trees\n",
    "\n",
    "all_dates = ['2022-03', '2022-04', '2022-05', '2022-06', '2022-07', \n",
    "             '2022-08', '2022-09', '2022-10', '2022-11', \n",
    "             '2023-01', '2023-02', '2023-03', '2023-04', '2023-05', '2023-06']\n",
    "\n",
    "for d in all_dates:\n",
    "    edit_tree(d, '21L', color_by_genes=all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2835db72",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
